python/ray/rllib/ddpg2/models.py

```diff
@@ -0,0 +1,408 @@
+        scope=scope if isinstance(scope, str) else scope.name)
+
+
+class ModelAndLoss(object):
+    """Holds the model and loss function.
+
+    Both graphs are necessary in order for the multi-gpu SGD implementation
+    to create towers on each device.
+    """
+
+    def __init__(
+            self, registry, dim_actions, low_action, high_action, config,
+            obs_t, act_t, rew_t, obs_tp1, done_mask, importance_weights):
+        # state outputs
+        with tf.variable_scope("s_func", reuse=True):
+            s_values = _build_s_network(
+                registry, obs_t, config)
+
+        with tf.variable_scope("target_s_func") as scope:
+            target_s_values = _build_s_network(
+                registry, obs_tp1, config)
+            self.target_s_func_vars = _scope_vars(scope.name)
+
+        # p network evaluation
+        with tf.variable_scope("p_func", reuse=True) as scope:
+            self.p_t = _build_p_network(registry, s_values, dim_actions, config)
+
+        # target p network evaluation
+        with tf.variable_scope("target_p_func") as scope:
+            self.p_tp1 = _build_p_network(registry, target_s_values, dim_actions, config)
+            self.target_p_func_vars = _scope_vars(scope.name)
+
+        # Action outputs
+        with tf.variable_scope("a_func", reuse=True):
+            deterministic_flag = tf.constant(value=False, dtype=tf.bool)
+            zero_eps = tf.constant(value=.0, dtype=tf.float32)
+            output_actions = _build_action_network(
+                self.p_t,
+                low_action,
+                high_action,
+                deterministic_flag,
+                zero_eps,
+                config["exploration_theta"],
+                config["exploration_sigma"])
+
+            output_actions_estimated = _build_action_network(
+                self.p_tp1,
+                low_action,
+                high_action,
+                deterministic_flag,
+                zero_eps,
+                config["exploration_theta"],
+                config["exploration_sigma"])
+
+        # q network evaluation
+        with tf.variable_scope("q_func", reuse=True) as scope:
+            self.q_t = _build_q_network(registry, s_values, act_t, config)
+        with tf.variable_scope("q_func", reuse=True) as scope:
+            self.q_tp0 = _build_q_network(registry, s_values, output_actions, config)
+
+        # target q network evalution
+        with tf.variable_scope("target_q_func") as scope:
+            self.q_tp1 = _build_q_network(registry, target_s_values, output_actions_estimated, config)
+            self.target_q_func_vars = _scope_vars(scope.name)
+
+        q_t_selected = tf.squeeze(self.q_t)
+
+        q_tp1_best = tf.squeeze(input=self.q_tp1)
+        q_tp1_best_masked = (1.0 - done_mask) * q_tp1_best
+
+        # compute RHS of bellman equation
+        q_t_selected_target = (rew_t + config["gamma"] ** config["n_step"] * q_tp1_best_masked)
```

Comment:
just checking: rew_t will have the discounted reward sum over the next n steps, correct?

Reply:
you got it. We use the `adjust_nsteps()` method defined in /rllib/dqn/dqn_evaluator.py